### ISSUE TITLE
`no-unresolved` is not aware of `exports` definition in `package.json`

### ISSUE THREAD

**Issue by ignatiusreza** [benmosher/eslint-plugin-import#1810]
From [this announcement](https://medium.com/@nodejs/announcing-core-node-js-support-for-ecmascript-modules-c5d6dc29b663)

> Package entry points and the package.json “exports” field
> 
> There are now two fields that can define entry points for a package: “main” and“exports”. The “main” field is supported in all versions of Node.js, but its capabilities are limited: it only defines the main entry point of the package. A new package.json field “exports” can also define the main entry point, along with subpaths. It also provides encapsulation, where only the paths explicitly defined in “exports” are available for importing. “exports” applies to both CommonJS and ES module packages, whether used via require or import.

Might need to be fix in [`resolve`](https://github.com/browserify/resolve/pull/224) package..

**Comment by ljharb** [comment]
Indeed, that’s where it needs to be fixed.

However, for back compact, you should have a file at the location the exports key would otherwise resolve to anyways - so even when it’s fixed in resolve, you’d still want it fixed now.

**Comment by plus-ultra-git** [comment]
When to fix?

**Comment by inlined** [comment]
Will this be resolved any time soon? We're going to have to yank `no-unresolved` from our Google code samples because it can no longer handle our SDKs.

**Comment by ljharb** [comment]
@inlined why would your Google SDKs be using `exports`, without `main`, in a way that's not backwards-compatible??

**Comment by inlined** [comment]
We _are_ using `main`. We also use `exports` for submodule paths so we can omit the `lib` folder generated by TypeScript.

For example, `./lib/v2/providers/https.js` is exported at `./v2/https`. See github.com/firebase/firebase-functions for more info.

**Comment by ljharb** [comment]
Right - but that's not backwards-compatible to pre-exports node. However, if you created an actual `./v2/https.js` file that re-exported that provider, then not only would your package be backwards-compatible, but you wouldn't be blocked on `resolve` getting `exports` support.

**Comment by inlined** [comment]
We control our execution environment and do not need to be backwards compatible to versions that don't support exports.

**Comment by ljharb** [comment]
@inlined i'm not "of the opinion that exports should not use renaming features", i'm of the opinion that you have a really simple workaround while you wait.

None of my "opinions" are delaying `exports` support - the work is difficult and nobody else is doing it, and I have limited time. It will be done eventually.

**Comment by ljharb** [comment]
@inlined if google wants to help make it happen faster, please feel free to visit https://github.com/browserify/resolve?sponsor=1

**Comment by danielweck** [comment]
Hi all, FYI I am successfully using this tiny ESLint import resolver to support ESM modules imports via package.json exports map:

https://gist.github.com/danielweck/cd63af8e9a8b3492abacc312af9f28fd

Duplicate issue? https://github.com/import-js/eslint-plugin-import/issues/1868

**Comment by bertho-zero** [comment]
I created a simple resolver that works thanks to [`enhanced-resolve`](https://github.com/webpack/enhanced-resolve):

**`resolver.js`**
```js
'use strict';

const path = require('path');
const { builtinModules } = require('module');
const enhancedResolve = require('enhanced-resolve');

const builtins = new Set(builtinModules);

function resolve(source, file, config) {
  if (builtins.has(source)) {
    return { found: true, path: null };
  }

  try {
    const resolver = config ? enhancedResolve.create.sync(config) : enhancedResolve.sync;
    const result = resolver(path.dirname(file), source);

    return { found: true, path: result };
  } catch (e) {
    return { found: false };
  }
};

module.exports = {
  interfaceVersion: 2,
  resolve
};
```

**Usage (`eslint.config.js`):**

```js
// without config
module.exports = {
  // ...
  settings: {
    'import/resolver': path.resolve(__dirname, './resolver')
  },
  // ...
};

// with `enhanced-resolve` config
module.exports = {
  // ...
  settings: {
    'import/resolver': {
      [path.resolve(__dirname, './resolver')]: {
        extensions: ['.js', '.ts']
      }
    }
  },
  // ...
};
```

**Comment by linguofeng** [comment]
Any updates?

**Comment by piranna** [comment]
@bertho-zero it looks similar to https://gist.github.com/danielweck/cd63af8e9a8b3492abacc312af9f28fd, can you create a `npm` package from it?

**Comment by cyco130** [comment]
I just published a resolver package that solves this issue using [resolve.exports](https://github.com/lukeed/resolve.exports): https://www.npmjs.com/package/eslint-import-resolver-exports

It's currently beta but seems to work for most of my projects. Feedback appreciated.

**Comment by DamienCassou** [comment]
> It's currently beta but seems to work for most of my projects. Feedback appreciated.

do you have examples that we could look at? I don't understand how to configure the resolver to get the same behavior as the one from the webpack resolver.

**Comment by cyco130** [comment]
> do you have examples that we could look at?

Currently what's in the readme is all I have. I think it's best to use this _in addition to_ other resolvers, as a fallback as it only supports `main`, `module`, `exports` and nothing else. The readme shows how to use it with TypeScript resolver for example. You should probably keep using Webpack's resolver too.

Check the [`resolve.exports` docs](https://github.com/lukeed/resolve.exports) for configuration options.

**Comment by JounQin** [comment]
You can use https://github.com/import-js/eslint-import-resolver-typescript which supports `exports` in `package.json` instead.

**Comment by ljharb** [comment]
That should only be a suggestion for TS users; the real solution here is for `resolve` to add support for `exports`.

**Comment by JounQin** [comment]
> That should only be a suggestion for TS users; the real solution here is for `resolve` to add support for `exports`.

Of course, I didn't close this issue. 🤣

**Comment by stagas** [comment]
> I just published a resolver package that solves this issue using [resolve.exports](https://github.com/lukeed/resolve.exports): https://www.npmjs.com/package/eslint-import-resolver-exports
> 
> It's currently beta but seems to work for most of my projects. Feedback appreciated.

Thank you @cyco130, that solved it for me.

**Comment by conartist6** [comment]
I'm proposing to fix this by writing a new resolver. I've been experimenting recently (with great success, I feel) in making code sync/async agnostic by making aggressive use of the [strategy pattern](https://gist.github.com/conartist6/abfb159d3c4494f4889e3d003558672c), which allows you to write functionally pure code which makes requests to and receives responses from a stateful core. Because that request/response mechanism is `yield`, the core is able to complete the request synchronously, or it may leave the generator suspended while it waits for async lookups to complete.

Would anyone be interested in collaborating me on a project like that if it would fix this issue once and for all?

**Comment by ljharb** [comment]
Fixing resolve would be of much larger impact than a new resolver.

**Comment by cyco130** [comment]
Hi @conartist6, I already have a working resolver I shared in [my comment](https://github.com/import-js/eslint-plugin-import/issues/1810#issuecomment-1188684050). I gladly accept contributions :)

**Comment by nia072011** [comment]
Why not use `createRequire(importer).resolve(importee,{paths:[]})` directly? In order to support configuring `extensions`, try resolve `importee.[ext]` and `importee/index.[ext]` again.

**Comment by ljharb** [comment]
@nia072011 that wouldn't work in older node versions, and still wouldn't support ESM. It's not a terrible workaround tho pending the `resolve` implementation.

**Comment by magnusriga** [comment]
Thanks @cyco130 et al.

Just so I have not misunderstood anything, I am getting an ESLint error when importing a TS module B into a JS file A, when B only uses `exports` in `package.json`. If I convert A into TypeScript, the error disappears (I am guessing that is because `eslint-import-resolver-typescript` supports the `exports` field).

To be clear, this works:
* Importing a JS module into a JS file
* Importing a JS module into a TS file
* Importing a TS module into a TS file

But, this does not work:
*  Importing a TS module into a JS file

Is that still the issue the rest of you see as well?

**Comment by ljharb** [comment]
@magnusriga this can be easily solved by making sure that the actual path to the file is present in the LHS of exports, so that pre-exports node (and resolve, and thus this plugin) can find it also;.

**Comment by magnusriga** [comment]
@ljharb Thank you for the reply. Do you mean I do not need the new [`eslint-import-resolver-exports`](https://www.npmjs.com/package/eslint-import-resolver-exports) package?

This is my `exports` field in `package.json` (I output `js` to `.dist` and types to `./types`):

![image](https://github.com/import-js/eslint-plugin-import/assets/38915578/fd920bb1-38d2-4466-b60c-953798c2ea89)

With the above, all imports work except when I try to import a TS module into a JS file (I am using `eslint-import-resolver-typescript).

**Comment by ljharb** [comment]
in that case you only have native ESM, which we don’t support yet.

**Comment by magnusriga** [comment]
I am not as knowledgable here, but what you are saying is that when I am using `import { B } from 'acme/ui'` I am using ESM style imports, and not CommonJS (`require`). And currently, a workaround (like the new [eslint-import-resolver-exports](https://www.npmjs.com/package/eslint-import-resolver-exports) package) is needed to allow TS files to import JS files when the latter is exported with `exports` (like in my screen shot above).

Is that correct? Thank you.

**Comment by ljharb** [comment]
The import style is irrelevant; the issue is what format the thing you're importing is using.

**Comment by magnusriga** [comment]
@ljharb Thanks for the clarification. Do you think `eslint-plugin-import` will support importing ESM components in the near future, or should we expect to rely on alternatives for the next few months?

**Comment by ljharb** [comment]
This plugin can't support `exports`, nor native ESM, until `resolve` does; I have no timeline for when `resolve` will support that, but I continue to hope it's "soon".

**Comment by conartist6** [comment]
What's it been so far,  three or four years?

**Comment by ljharb** [comment]
Yup. Doing unpaid volunteer work doesn't create a lot of free time.

**Comment by conartist6** [comment]
I'm just gonna build it,  unless anyone objects. Now is a good moment for
me.

**Comment by ljharb** [comment]
You've always been welcome to, but unless it's in the `resolve` package it won't have the same benefit, I'm afraid.

**Comment by conartist6** [comment]
Could you clarify what you mean?

I think of this as a case that demands competive compatibility, but so long
as I'm doing a full rewrite I will have little desire to release my work
under their name (even if it were possible).

**Comment by ljharb** [comment]
it is, as I’m the sole maintainer of `resolve`, but a full rewrite isn’t needed anyways.

**Comment by conartist6** [comment]
I'd rather not build it in a way that all the important logic is duplicated between the sync and async pathways. My aim would be define the resolution algorithm once and only once, and generator functions are the means to do it.

**Comment by conartist6** [comment]
There aren't any versions of node that support `exports` but don't support generators, right? In other words, old versions of node would continue to use your package.

My primary concern would be ensuring that if users who are coding the most modern and correct style have readily available tools that can help them.

**Comment by conartist6** [comment]
ESLlint 4.19.1, now six years old, requires node `>4`, which already supported ES6. Old code that actually resolves modules won't know about `exports` anyway, so as I say for those users there would be no need to update this package at all

**Comment by ljharb** [comment]
The resolution needs to be done regardless of the current node version to be usable here.

**Comment by conartist6** [comment]
I'm just trying to figure out where the requirement for the resolver to be implemented in es3 is coming from

**Comment by ljharb** [comment]
From the node versions it needs to support. Generators are garbage anyways and should be avoided, imo.

**Comment by conartist6** [comment]
Could you clarify: why do really old node versions need to be able to resolve `exports` though?

**Comment by ljharb** [comment]
The use cases for resolve are far more plentiful than for eslint. resolve needs to resolve, in any node version, a path *as if* it would be resolved in any other node version. A pre-exports node may want to emulate “exports”. A post-exports node may want to emulate pre-exports. A modern node may simply want to know what a pre-exports node user would see.

**Comment by piranna** [comment]
I think the sane solution here is to don't worry at all about any unmaintained Node.js version, it's said, don't provide support for anything older than Node.js v18.

**Comment by conartist6** [comment]
I really don't buy that the only way to have good compatibility is if there is a single implementation of the algorithm. Tons of projects are already using `enhanced-resolve` anyway, and putting a new algorithm out in ESLint will be an amazing way to safely get feedback on whether there is behavioral parity.

**Comment by ljharb** [comment]
@piranna ”sane” isn’t an appropriate word to be using in this context; and almost all of my packages support down to node 0.4, and will continue to do so for the foreseeable future. Platform support has no correlation to what a package should support.

**Comment by ljharb** [comment]
@conartist6 thats fine, your agreement isn’t required.

**Comment by conartist6** [comment]
I think this is great because now we're unblocked. It's clear from this conversation that my beliefs about the problem are different enough from yours that I should build a new package, and that it will be clearly differentiated in the marketplace.

**Comment by conartist6** [comment]
Everyone who just wants to use the new stuff can stop bugging you, which is a win for you as well.

**Comment by ljharb** [comment]
Go for it. As you said tho, there’s already packages that do this, so i don’t think there’s much value in yet another one; the value is in resolve itself doing it.

**Comment by conartist6** [comment]
I would agree that all that was off-topic if this were a conversation on the `resolve` package, but as this is `eslint-plugin-import` parts of that conversation seem quite relevant.

To recap the now-hidden conversation:
- @ljharb  is the sole maintainer of the `resolve` package
- @ljharb's mission for the resolve package is to support node `>= 0.4.0`, meaning he can only work in es3
- ESLint has required node `>= 4.0.0` for the last six years (at least)
- `resolve` favors speed and compatibility over readability and correctness

I believe that in sum this means that `eslint-plugin-import` should seek out a resolve implementation which is better aligned with its needs and the needs of its users.

**Comment by ljharb** [comment]
It won’t; using resolve - the most common implementation - is what is most aligned with it and its users needs.

**Comment by conartist6** [comment]
I would appreciate if you would allow my condensed argument to stand, for the record. Please?

**Comment by ljharb** [comment]
Your summary would be fine; your conclusion would only distract.

**Comment by icholy** [comment]
Wow .. @ljharb you're a loser

**Comment by conartist6** [comment]
The comment may be deleted now, but I'm still calling it out. That was rude and inappropriate. We can compete with and disagree with other people and still respect them.

**Comment by ljharb** [comment]
@conartist6 no comments were deleted afaict, so I’m not sure what you’re referring to.

**Comment by conartist6** [comment]
Not deleted by you, but it was emailed to anyone who subscribes to this thread.

**Comment by ljharb** [comment]
ah, yes, i found it in my email :-) some people have nothing better to do, i guess.

**Comment by conartist6** [comment]
I feel that you have abused your position as a moderator to protect your image in the face of reasoned dissent, so I can't say I'm that thrilled with you at the moment.

**Comment by conartist6** [comment]
Is there an ombudsman or another moderator who could assist in untangling this?

**Comment by conartist6** [comment]
I'm not wedded to any particular outcome, but it's unacceptable to me that all discussion on the topic should be summarily silenced.

**Comment by ljharb** [comment]
It’s just me, and what’s acceptable is only up to collaborators. I’m going to lock the thread for a bit.

**Comment by Eunknight** [comment]
causes error
`/node_modules/@tailwindcss/vite/package.json` 
```javascript
"exports": {
  ".": {
    "import": "./dist/index.mjs",
    "types": "./dist/index.d.mts"
  }
},
``` 
---
not error 1, when
```javascript
"exports": {
  ".": "./dist/index.mjs"
  }
``` 
---
not error 2,
AI answer

```
"exports": {
  ".": {
    "import": {
      "types": "./dist/index.d.mts",
      "default": "./dist/index.mjs"
    },
    "default": "./dist/index.mjs"
  }
}
```
---
and ai suggest add
```
"main": "./dist/index.mjs",
"types": "./dist/index.d.mts",
```

full code
```
"main": "./dist/index.mjs",
"types": "./dist/index.d.mts",
"exports": {
  ".": {
    "import": {
      "types": "./dist/index.d.mts",
      "default": "./dist/index.mjs"
    },
    "default": "./dist/index.mjs"
  }
}
```

**Comment by fhanggi** [comment]
Would this be a good workaround to take care of the eslint import/no-unresolved errors caused by this resolver issue or will this hide something?

adding this to eslint.config.mjs under settings

node: {
  extensions: [".js", ".jsx"],
  exportsCategory: "conditions"
}
